File: JavaScript/206.反转链表.js
/*
 * @lc app=leetcode.cn id=206 lang=javascript
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
/* 
思路：双指针互相☞
这种基础的破题过段时间就得刷一下加深下记忆
*/
/* 
Accepted
28/28 cases passed (96 ms)
Your runtime beats 36.69 % of javascript submissions
Your memory usage beats 45.02 % of javascript submissions (39.7 MB)
*/
var reverseList = function (head) {
  let p = head
  let l = null
  while (p) {
    temp = p.next
    p.next = l
    l = p
    p = temp
  }
  return l
};
// @lc code=end

File: Python3/206.反转链表.py
#
# @lc app=leetcode.cn id=206 lang=python3
#
# [206] 反转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
"""
双指针与python的骚操作
Accepted
28/28 cases passed (48 ms)
Your memory usage beats 61.16 % of python3 submissions (15.5 MB
Your runtime beats 11.49 % of python3 submissions
"""


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        prev, cur = head, None
        while prev:
            prev.next, prev, cur = cur, prev.next, prev
        return cur
# @lc code=end
